src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteProperties.String.cs

```diff
@@ -9,6 +9,353 @@ namespace System.Text.Json
+        /// <remarks>
+        /// The property name is escaped before writing.
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the specified property name is too large.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WritePropertyName(string propertyName)
+            => WritePropertyName(propertyName.AsSpan());
+
+        /// <summary>
+        /// Writes the property name (as a JSON string) as the first part of a name/value pair of a JSON object.
+        /// </summary>
+        /// <param name="propertyName">The property name of the JSON object to be transcoded and written as UTF-8.</param>
+        /// <remarks>
+        /// The property name is escaped before writing.
+        /// </remarks>
+        /// <exception cref="ArgumentException">
+        /// Thrown when the specified property name is too large.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if this would result in an invalid JSON to be written (while validation is enabled).
+        /// </exception>
+        public void WritePropertyName(ReadOnlySpan<char> propertyName)
+        {
+            JsonWriterHelper.ValidateProperty(propertyName);
+
+            int propertyIdx = JsonWriterHelper.NeedsEscaping(propertyName);
+
+            Debug.Assert(propertyIdx >= -1 && propertyIdx < propertyName.Length && propertyIdx < int.MaxValue / 2);
+
+            if (propertyIdx != -1)
+            {
+                WriteStringEscapeProperty(propertyName, propertyIdx);
+            }
+            else
+            {
+                WriteStringByOptionsPropertyName(propertyName);
+            }
+            _currentDepth &= JsonConstants.RemoveFlagsBitMask;
+            _tokenType = JsonTokenType.PropertyName;
+        }
+
+        private void WriteStringEscapeProperty(ReadOnlySpan<char> propertyName, int firstEscapeIndexProp)
+        {
+            Debug.Assert(int.MaxValue / JsonConstants.MaxExpansionFactorWhileEscaping >= propertyName.Length);
+
+            char[] propertyArray = null;
+
+            if (firstEscapeIndexProp != -1)
+            {
+                int length = JsonWriterHelper.GetMaxEscapedLength(propertyName.Length, firstEscapeIndexProp);
+
+                Span<char> escapedPropertyName;
+                if (length > JsonConstants.StackallocThreshold)
+                {
+                    propertyArray = ArrayPool<char>.Shared.Rent(length);
+                    escapedPropertyName = propertyArray;
+                }
+                else
+                {
+                    // Cannot create a span directly since it gets assigned to parameter and passed down.
```

Comment:
Are we not able to work around this now with readonly members?

Reply:
Copy-pasted from existing code, so will have to fix this across the board. Out-of-scope for this PR.